> Can anyone tell me why the twister talent failed with the same code it passed with for the last week?

See the errors:

/__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function]
   74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action)
      |            ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in https://github.com/zephyrproject-rtos/zephyr/pull/77110

--- a/boards/ambiq/apollo4p_evb/apollo4p_evb-pinctrl.dtsi
+++ b/boards/ambiq/apollo4p_evb/apollo4p_evb-pinctrl.dtsi
@@ -17,6 +17,12 @@
 			input-enable;
 		};
 	};
+	adc0_default: adc0_default{
+		group1 {
+			pinmux = <ADCSE4_P15>, <ADCSE7_P12>;
+			drive-strength = "0.1";
+		};
+	};
 	i2c0_default: i2c0_default {
 		group1 {
 			pinmux = <M0SCL_P5>, <M0SDAWIR3_P6>;
@@ -81,7 +87,6 @@
 			bias-pull-up;
 		};
 	};
-
 	spi0_default: spi0_default {
 		group1 {
 			pinmux = <M0SCK_P5>, <M0MISO_P7>, <M0MOSI_P6>;

--- a/boards/ambiq/apollo4p_evb/apollo4p_evb.dts
+++ b/boards/ambiq/apollo4p_evb/apollo4p_evb.dts
@@ -63,6 +63,13 @@
 	status = "okay";
 };
 
+&adc0 {
+compatible = "ambiq,adc";
+	pinctrl-0 = <&adc0_default>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
 &counter0 {
 	status = "okay";
 };

--- a/boards/ambiq/apollo4p_evb/apollo4p_evb.yaml
+++ b/boards/ambiq/apollo4p_evb/apollo4p_evb.yaml
@@ -15,6 +15,7 @@ supported:
   - spi
   - i2c
   - rtc
+  - adc
   - hwinfo
 testing:
   ignore_tags:

--- a/dts/arm/ambiq/ambiq_apollo4p.dtsi
+++ b/dts/arm/ambiq/ambiq_apollo4p.dtsi
@@ -3,6 +3,7 @@
 #include <arm/armv7-m.dtsi>
 #include <mem.h>
 #include <freq.h>
+#include <zephyr/dt-bindings/adc/adc.h>
 #include <zephyr/dt-bindings/i2c/i2c.h>
 #include <zephyr/dt-bindings/gpio/gpio.h>
 
@@ -286,6 +287,17 @@
 			interrupts = <13 0>;
 			status = "disabled";
 			ambiq,pwrcfg = <&pwrcfg 0x4 0x100>;
+		};
+
+		adc0: adc@40038000 {
+			reg = <0x40038000 0x400>;
+			interrupts = <19 0>;
+			interrupt-names = "ADC";
+			channel-count = <10>;
+			internal-vref-mv = <1190>;
+			status = "disabled";
+			#io-channel-cells = <1>;
+			ambiq,pwrcfg = <&pwrcfg 0x4 0x2000>;
 		};
 
 		mspi0: spi@40060000 {

--- a/modules/hal_ambiq/Kconfig
+++ b/modules/hal_ambiq/Kconfig
@@ -10,6 +10,11 @@ config AMBIQ_HAL
 
 if AMBIQ_HAL
 
+config AMBIQ_HAL_USE_ADC
+	bool
+	help
+	  Use the ADC driver from Ambiq HAL
+
 config AMBIQ_HAL_USE_GPIO
 	bool
 	help

--- a/samples/drivers/adc/adc_dt/boards/apollo4p_evb.overlay
+++ b/samples/drivers/adc/adc_dt/boards/apollo4p_evb.overlay
@@ -1,0 +1,34 @@
+/*
+ * Copyright (c) 2024 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/ {
+	zephyr,user {
+		io-channels = <&adc0 4>, <&adc0 7>;
+	};
+};
+
+&adc0 {
+	status = "okay";
+	interrupt-parent = <&nvic>;
+	interrupts = <19 0>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	channel@4 {
+		reg = <4>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,resolution = <12>;
+	};
+
+	channel@7 {
+		reg = <7>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,resolution = <12>;
+	};
+};

--- a/tests/drivers/adc/adc_api/boards/apollo4p_evb.overlay
+++ b/tests/drivers/adc/adc_api/boards/apollo4p_evb.overlay
@@ -1,0 +1,32 @@
+/*
+ * Copyright (c) 2024 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/ {
+	zephyr,user {
+		io-channels = <&adc0 4>, <&adc0 7>;
+	};
+};
+
+&adc0 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	channel@4 {
+		reg = <4>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,resolution = <12>;
+	};
+
+	channel@7 {
+		reg = <7>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,resolution = <12>;
+	};
+};

